R/input.R

```diff
@@ -0,0 +1,268 @@
+##' Control for query parameters.
+##'
+##' @title Control for query parameters
+##'
+##' @param ... Named arguments representing accepted parameters.  The
+##'   value of each must be a type.
+##'
+##' @param .parameters A list of named parameters to accept, instead
+##'   of using \code{...} - this interface is considerably easier to
+##'   program against if building an API programmatically, avoiding
+##'   the use of \code{\link{do.call}}.
+##'
+##' @export
+##'
+##' @examples
+##' pkgapi::pkgapi_input_query(number = "integer")
+pkgapi_input_query <- function(..., .parameters = list(...)) {
+  assert_named(.parameters, TRUE)
+  nms <- names(.parameters)
+
+  types <- c("logical", "numeric", "integer", "string")
+
+  for (i in seq_along(.parameters)) {
+    match_value(.parameters[[i]], types,
+                sprintf("The 'type' of parameter %s", nms[[i]]))
+  }
+  types <- vcapply(.parameters, identity, USE.NAMES = FALSE)
+  unname(Map(pkgapi_input, nms, types, MoreArgs = list(where = "query")))
+}
+
+
+##' Control for body parameters.  This might change.  There are
+##' several types of HTTP bodies that we want to consider here - the
+##' primary ones are a body uploaded in binary, the other is a json
+##' object.  In the latter we want to validate the body against a
+##' schema (at least if validation is used).  In future we might also
+##' support a form input here too.
+##'
+##' @title Control for body parameters
+##'
+##' @param name Name of the parameter
+##'
+##' @export
+##' @rdname pkgapi_input_body
+pkgapi_input_body_binary <- function(name) {
+  assert_scalar_character(name)
+  pkgapi_input(name, "binary", "body", assert_raw,
+               content_type = "application/octet-stream")
+}
+
+
+##' @inheritParams pkgapi_returning_json
+##' @export
+##' @rdname pkgapi_input_body
+pkgapi_input_body_json <- function(name, schema, root) {
+  assert_scalar_character(name)
+  validator <- pkgapi_validator(schema, schema_root(root), query = NULL)
+  pkgapi_input(name, "json", "body", validator,
+               content_type = "application/json")
+}
+
+
+pkgapi_input_path <- function(path) {
+  data <- parse_plumber_path(path)
+  if (is.null(data)) {
+    return(NULL)
+  }
+  apply(parse_plumber_path(path), 1, function(x)
+    pkgapi_input(x[[1]], x[[2]], "path"))
+}
+
+
+pkgapi_input <- function(name, type, where, validator = NULL, ...) {
+  res <- list(name = name, type = type, where = where, validator = validator,
+              ...)
+  if (is.null(validator)) {
+    res$validator <- pkgapi_input_validator_basic(type)
+  }
+  class(res) <- "pkgapi_input"
+  res
+}
+
+
+pkgapi_inputs_init <- function(path, inputs_query, inputs_body, args) {
```

Comment:
I'm getting pretty confused around here. There is just lots of moving parts I'm struggling to keep track of in my head. Lots of closures being passed around and these arguments are all lists with various properties which it is hard to keep track of without types or documentation.

We talked about this in person moving to using some more objects here to hopefully tidy up some of the tricky bits.

Reply:
ok, as discussed, this is now overhauled and considerably better for your comments